File: S2013/DPL/HW4/T9Spelling/T9Spelling.hs
module Main where
import Data.List
-- Solve the Google Code Jam problem "T9 Spelling":
--   http://code.google.com/codejam/contest/351101/dashboard#s=p2

main = do
    count <- getLine
    lines <- getLines (read count :: Int)
    mapM_ putStrLn $ addCaseLabels $ map tNine lines

-- mapM_ is "for each"
--   Inside a "do" block, it takes a function and a list and calls
--   the function on each item in the list as if the calls were 
--   directly in the "do" block.

getLines 0 = return []
getLines n = do
    x  <- getLine
    xs <- getLines (n - 1)
    return $ x : xs

addCaseLabels xs = map addCase $ zip [1..] xs
    where addCase (nn, ss) = "Case #" ++ (show nn) ++ ": " ++ ss


tNine x = tNinetoNum (tNineRec x []) []

tNineRec :: [Char] -> [Char] -> [Char]
tNineRec (x:[]) acc = acc ++ [x]
tNineRec (x:y:z) acc 
    | map (x `elem`) ["abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"," "] == map (y `elem`) ["abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"," "] = tNineRec (y:z) (acc ++ [x] ++ ".")
    | otherwise = tNineRec (y:z) (acc ++ [x])
 
tNinetoNum [] acc = acc
tNinetoNum (x:y) acc
    | x == 'a' = tNinetoNum y $ acc ++ "2"
    | x == 'b' = tNinetoNum y $ acc ++ "22"
    | x == 'c' = tNinetoNum y $ acc ++ "222"
    | x == 'd' = tNinetoNum y $ acc ++ "3"
    | x == 'e' = tNinetoNum y $ acc ++ "33"
    | x == 'f' = tNinetoNum y $ acc ++ "333"
    | x == 'g' = tNinetoNum y $ acc ++ "4"
    | x == 'h' = tNinetoNum y $ acc ++ "44"
    | x == 'i' = tNinetoNum y $ acc ++ "444"
    | x == 'j' = tNinetoNum y $ acc ++ "5"
    | x == 'k' = tNinetoNum y $ acc ++ "55"
    | x == 'l' = tNinetoNum y $ acc ++ "555"
    | x == 'm' = tNinetoNum y $ acc ++ "6"
    | x == 'n' = tNinetoNum y $ acc ++ "66"
    | x == 'o' = tNinetoNum y $ acc ++ "666"
    | x == 'p' = tNinetoNum y $ acc ++ "7"
    | x == 'q' = tNinetoNum y $ acc ++ "77"
    | x == 'r' = tNinetoNum y $ acc ++ "777"
    | x == 's' = tNinetoNum y $ acc ++ "7777"
    | x == 't' = tNinetoNum y $ acc ++ "8"
    | x == 'u' = tNinetoNum y $ acc ++ "88"
    | x == 'v' = tNinetoNum y $ acc ++ "888"
    | x == 'w' = tNinetoNum y $ acc ++ "9"
    | x == 'x' = tNinetoNum y $ acc ++ "99"
    | x == 'y' = tNinetoNum y $ acc ++ "999"
    | x == 'z' = tNinetoNum y $ acc ++ "9999"
    | x == ' ' = tNinetoNum y $ acc ++ "0"
    | x == '.' = tNinetoNum y $ acc ++ " "

File: S2013/DPL/HW9/CodeJam/t9_spelling.rb
#!/usr/bin/env ruby

# Please solve the T9 Spelling problem from Google Code Jam:
#   http://code.google.com/codejam/contest/351101/dashboard#s=p2
#
# The test harness expects you to open the file named in ARGV[0]
# yourself rather than accepting the input data on STDIN.


class T9
    def read(thefile)
        file = File.new(thefile)
        @counter = file.gets.to_i
        @cases = []
        file.each do |row|
            @cases << row.delete("\n")
        end
    end
    def counter
        @counter
    end
    def cases
        @cases
    end
    def addpause(s)
        s = s.gsub(/([abc]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([def]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([ghi]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([jkl]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([mno]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([pqrs]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([tuv]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([wxyz]){2,}/) {|s| s.split('').join('.')}
        s = s.gsub(/([ ]){2,}/) {|s| s.split('').join('.')}
    end
    def converttonum(s)
        s = s.gsub(/a/,'2')
        s = s.gsub(/b/,'22')
        s = s.gsub(/c/,'222')
        s = s.gsub(/d/,'3')
        s = s.gsub(/e/,'33')
        s = s.gsub(/f/,'333')
        s = s.gsub(/g/,'4')
        s = s.gsub(/h/,'44')
        s = s.gsub(/i/,'444')
        s = s.gsub(/j/,'5')
        s = s.gsub(/k/,'55')
        s = s.gsub(/l/,'555')
        s = s.gsub(/m/,'6')
        s = s.gsub(/n/,'66')
        s = s.gsub(/o/,'666')
        s = s.gsub(/p/,'7')
        s = s.gsub(/q/,'77')
        s = s.gsub(/r/,'777')
        s = s.gsub(/s/,'7777')
        s = s.gsub(/t/,'8')
        s = s.gsub(/u/,'88')
        s = s.gsub(/v/,'888')
        s = s.gsub(/w/,'9')
        s = s.gsub(/x/,'99')
        s = s.gsub(/y/,'999')
        s = s.gsub(/z/,'9999')
        s = s.gsub(/ /,'0')        
        s = s.gsub(/\./,' ')
    end
end
a = T9.new
a.read(ARGV[0])
i = 0
while i < a.counter
    puts "Case ##{i+1}: #{a.converttonum(a.addpause(a.cases[i]))}"
    i = i + 1
end